plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java

```diff
@@ -0,0 +1,213 @@
+import org.eclipse.m2e.core.MavenPlugin;
+import org.eclipse.m2e.core.project.IMavenProjectFacade;
+import org.eclipse.wst.server.core.IModule;
+import org.eclipse.wst.server.core.IRuntime;
+import org.eclipse.wst.server.core.IServer;
+import org.eclipse.wst.server.core.IServerListener;
+import org.eclipse.wst.server.core.ServerEvent;
+import org.eclipse.wst.server.core.ServerUtil;
+
+import com.google.cloud.tools.eclipse.appengine.localserver.Activator;
+import com.google.cloud.tools.eclipse.appengine.localserver.GCloudCommandDelegate;
+
+/**
+ * Cloud SDK server's launch configuration delegate.
+ */
+public class CloudSdkLaunchConfigurationDelegate extends AbstractJavaLaunchConfigurationDelegate {
+  private IServerListener debugServerListener;
+  private static final String DEBUGGER_HOST = "localhost";
+
+  @Override
+  public void launch(ILaunchConfiguration configuration, String mode, ILaunch launch, IProgressMonitor monitor)
+      throws CoreException {
+    final IServer server = ServerUtil.getServer(configuration);
+    if (server == null) {
+      return;
+    }
+
+    final IModule[] modules = server.createWorkingCopy().getModules();
+    if (modules == null || modules.length == 0) {
+      return;
+    }
+
+    String runnables = getRunnable(modules[0].getProject(), monitor);
+    IRuntime runtime = server.getRuntime();
+    if (runtime == null) {
+      return;
+    }
+    IPath sdkLocation = runtime.getLocation();
+
+    CloudSdkServerBehaviour sb = (CloudSdkServerBehaviour) server.loadAdapter(CloudSdkServerBehaviour.class, null);
+    sb.setupLaunch(mode);
+
+    final CloudSdkServer cloudSdkServer = CloudSdkServer.getCloudSdkServer(server);
+    int debugPort = -1;
+    if (mode.equals(ILaunchManager.DEBUG_MODE)) {
+      final int port = getDebugPort();
+      debugPort = port;
+
+      // Create server listener so that when server has started we create a
+      // remote debugger
+      // and attach it to the server
+      debugServerListener = new IServerListener() {
+        @Override
+        public void serverChanged(ServerEvent event) {
+          if (event != null) {
+            if (event.getState() == IServer.STATE_STARTED) {
+              try {
+                Thread.sleep(500);
+                runDebugTarget(cloudSdkServer, modules[0].getProject().getName(), port);
+              } catch (InterruptedException | CoreException e) {
+                Activator.logError(e);
+              }
+            } else if (event.getState() == IServer.STATE_STOPPING) {
+              removeRemoteDebuggerListener(server);
+            }
+
+          }
+        }
+      };
+      server.addServerListener(debugServerListener);
+    }
+
+    String commands = GCloudCommandDelegate.createAppRunCommand(sdkLocation.toOSString(), runnables, mode,
+        cloudSdkServer.getApiHost(), cloudSdkServer.getApiPort(), debugPort);
+    Activator.logInfo("Executing: " + commands);
+
+    Process p = null;
```

Comment:
try not to split variables across try blocks. In general more code should go into a try, not less